### PR TITLE
Fix `vyos.vyos.vyos_firewall_rules` `state: replaced` to match documentation

### LINKED ISSUE #201: `vyos.vyos.vyos_firewall_rules` cannot replace a single rule set
##### SUMMARY

While the documentation (via the examples section) indicates that `state: replaced` replaces the device configuration of the **listed** firewall rules, in reality `vyos.vyos.vyos_firewall_rules` will try to delete other rules not listed in the configuration.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

`vyos.vyos.vyos_firewall_rules`

##### ANSIBLE VERSION

```paste below
ansible 2.10.12
  config file = /home/sdwilsh/ansible-playbooks/ansible.cfg
  configured module search path = ['/home/sdwilsh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.10 (default, Jun  2 2021, 10:49:15) [GCC 9.4.0]
```

##### COLLECTION VERSION

```paste below
# /home/sdwilsh/.ansible/collections/ansible_collections
Collection Version
---------- -------
vyos.vyos  2.5.0  
[WARNING]: Collection at '/home/sdwilsh/ansible-collection-dev/ansible_collections/vyos/vyos' does not have a valid version set, falling back to '*'. Found version: 'None'

# /home/sdwilsh/ansible-collection-dev/ansible_collections
Collection Version
---------- -------
vyos.vyos  * 
```

##### CONFIGURATION

```paste below
COLLECTIONS_PATHS(/home/sdwilsh/ansible-playbooks/ansible.cfg) = ['/home/sdwilsh/ansible-collection-dev', '/home/sdwilsh/.ansible/collections', '/usr/share/ansible/collections']
DEFAULT_HOST_LIST(/home/sdwilsh/ansible-playbooks/ansible.cfg) = ['/home/sdwilsh/ansible-playbooks/prod-inventory']
DEFAULT_REMOTE_USER(/home/sdwilsh/ansible-playbooks/ansible.cfg) = ansible
DEFAULT_ROLES_PATH(/home/sdwilsh/ansible-playbooks/ansible.cfg) = ['/home/sdwilsh/ansible-playbooks/roles', '/home/sdwilsh/.ansible/roles']
DEFAULT_SCP_IF_SSH(/home/sdwilsh/ansible-playbooks/ansible.cfg) = true
DEFAULT_VAULT_PASSWORD_FILE(/home/sdwilsh/ansible-playbooks/ansible.cfg) = /home/sdwilsh/ansible-playbooks/.ansible_vault_password
INTERPRETER_PYTHON(/home/sdwilsh/ansible-playbooks/ansible.cfg) = auto
PERSISTENT_COMMAND_TIMEOUT(/home/sdwilsh/ansible-playbooks/ansible.cfg) = 180
```

##### OS / ENVIRONMENT

I'm targeting vyos 1.3 rc5, but given the commands being run, I don't think it matters.

##### STEPS TO REPRODUCE

```yaml
- name: Setup two rules for testing issue
  vyos.vyos.vyos_firewall_rules:
    config:
    - afi: ipv4
      rule_sets:
      - name: TEST
        default_action: drop
        rules:
        - number: 1
          action: accept
          state:
            established: true
            related: true
        - number: 2
          action: drop
          state:
            invalid: true
        - number: 3
          action: accept
          protocol: icmp
          icmp:
            type_name: any
        - number: 4
          action: accept
          protocol: vrrp
      - name: UNTOUCHED
        default_action: drop
        rules:
        - number: 1
          action: accept
          state:
            established: true
            related: true
        - number: 2
          action: drop
          state:
            invalid: true
    state: merged
- name: Replace device configurations of listed firewall rules with provided configurations
  vyos.vyos.vyos_firewall_rules:
    config:
    - afi: ipv4
      rule_sets:
      - name: TEST
        default_action: drop
        rules:
        - number: 1
          action: accept
          state:
            established: true
            related: true
        - number: 2
          action: drop
          state:
            invalid: true
        - number: 3
          action: accept
          protocol: vrrp
        - number: 4
          action: accept
          protocol: icmp
          icmp:
            type_name: any
    state: replaced
```

##### EXPECTED RESULTS

Firewall rule `UNTOUCHED` does not get changed, and `TEST` rules 3 and 4 change their order.

##### ACTUAL RESULTS

`UNTOUCHED`, and any additional firewall rules are deleted (if they aren't in use, at least).

I skipped the setup step below, since it has a ton of output due to a large number of firewall rules for various zones on the box I ran it on (one of which you can see in the error message was being deleted, and why the command failed).

```paste below
TASK [Replace device configurations of listed firewall rules with provided configurations] ***
task path: /home/sdwilsh/ansible-playbooks/test.yml:44
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
redirecting (type: terminal) ansible.builtin.vyos to vyos.vyos.vyos
redirecting (type: cliconf) ansible.builtin.vyos to vyos.vyos.vyos
redirecting (type: action) vyos.vyos.vyos_firewall_rules to vyos.vyos.vyos
redirecting (type: action) vyos.vyos.vyos_firewall_rules to vyos.vyos.vyos
<10.10.0.48> ESTABLISH LOCAL CONNECTION FOR USER: sdwilsh
<10.10.0.48> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb `"&& mkdir "` echo /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834 `" && echo ansible-tmp-1629729912.218707-565503-246921439531834="` echo /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834 `" ) && sleep 0'
Using module file /home/sdwilsh/ansible-collection-dev/ansible_collections/vyos/vyos/plugins/modules/vyos_firewall_rules.py
<10.10.0.48> PUT /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/tmpd9qnzsim TO /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py
<10.10.0.48> EXEC /bin/sh -c 'chmod u+x /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/ /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py && sleep 0'
<10.10.0.48> EXEC /bin/sh -c '/usr/bin/python3 /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py && sleep 0'
<10.10.0.48> EXEC /bin/sh -c 'rm -f -r /home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py", line 102, in <module>
    _ansiballz_main()
  File "/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.vyos.vyos.plugins.modules.vyos_firewall_rules', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib/python3.8/runpy.py", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/modules/vyos_firewall_rules.py", line 1514, in <module>
  File "/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/modules/vyos_firewall_rules.py", line 1509, in main
  File "/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py", line 87, in execute_module
  File "/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
ansible.module_utils.connection.ConnectionError: commit failed: E-v6]] failed
[ firewall ipv6-name HOME-from-MAN-v6 ]
Firewall configuration error: Cannot delete rule set "HOME-from-MAN-v6" (still in use)



[[firewall ipv6-name HOME-from-MAN-v6]] failed
Commit failed
[edit]

ansible@maccabeus01.hogs.tswn.us# 
fatal: [maccabeus01.hogs.tswn.us]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/sdwilsh/.ansible/tmp/ansible-local-565013_tnwu0sb/ansible-tmp-1629729912.218707-565503-246921439531834/AnsiballZ_vyos_firewall_rules.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.vyos.vyos.plugins.modules.vyos_firewall_rules', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/modules/vyos_firewall_rules.py\", line 1514, in <module>\n  File \"/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/modules/vyos_firewall_rules.py\", line 1509, in main\n  File \"/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible_collections/vyos/vyos/plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py\", line 87, in execute_module\n  File \"/tmp/ansible_vyos.vyos.vyos_firewall_rules_payload_1ensys4h/ansible_vyos.vyos.vyos_firewall_rules_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: commit failed: E-v6]] failed\r\n[ firewall ipv6-name HOME-from-MAN-v6 ]\r\nFirewall configuration error: Cannot delete rule set \"HOME-from-MAN-v6\" (still in use)\r\n\r\n\r\n\r\n[[firewall ipv6-name HOME-from-MAN-v6]] failed\r\nCommit failed\r\n[edit]\r\r\nansible@maccabeus01.hogs.tswn.us# \n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

### PR DESCRIPTION
**Pull request by sdwilsh**
##### SUMMARY
`vyos.vyos.vyos_firewall_rules` should only try to change listed firewall rules, as documented, when the state is set to `replaced`.  As currently implemented (prior to this PR), it better matches what `overridden` is meant to do.

Fixes #201

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
`vyos.vyos.vyos_firewall_rules`

##### ADDITIONAL INFORMATION

- [x] Cleanup and document existing code for clarity
- [x] Add a failing idempotent test
- [x] Add a failing change test
- [x] Fix failing tests
- [x] Add change fragment

### CODE REVIEW

File: plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py [hunk 1]
```diff
@@ -825,29 +842,29 @@ def _add_src_or_dest(self, attr, w, h, cmd, opr=True):
                             )
         return commands
 
-    def search_r_sets_in_have(self, have, w_name, type="rule_sets"):
+    def search_r_sets_in_target(self, target, w_name, type="rule_sets"):
```
Comment (GomathiselviS) [inline plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py:845]: Why is this change needed? I see the change in the name of the function and arguments. Am I missing something?

Comment (sdwilsh) [reply plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py:845]: I tried to explain this change in the commit message where I made it (3afa41de0666def812e275e4941a696eca999542), but GitHub doesn't make those super discoverable, alas.

The change was largely because I kept getting very confused because methods that would call this would pass in their have or their want as the `have` for this method, and the swapping was making it hard to reason about.  I choose `target` because it is independent of `have` and `want`.

**Comment by sdwilsh** [comment]
In terms of the CI errors, the `black` failure was me just misreading the logs last night and I can fix that this evening.  The lint errors all are in files I have not touched, so they appear to be in the `main` branch today:
```
ERROR: Found 7 pylint issue(s) which need to be resolved:
ERROR: plugins/cliconf/vyos.py:171:4: arguments-renamed: Parameter 'newline' has been renamed to 'output' in overridden 'Cliconf.get' method
ERROR: plugins/cliconf/vyos.py:171:4: arguments-renamed: Parameter 'output' has been renamed to 'newline' in overridden 'Cliconf.get' method
ERROR: tests/unit/mock/loader.py:49:4: arguments-renamed: Parameter 'file_name' has been renamed to 'path' in overridden 'DictDataLoader._get_file_contents' method
ERROR: tests/unit/modules/network/vyos/test_vyos_interface.py:89:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosInterfaceModule.load_fixtures' method
ERROR: tests/unit/modules/network/vyos/test_vyos_ping.py:52:27: use-maxsplit-arg: Use str(command).split(' | ', maxsplit=1)[0] instead
ERROR: tests/unit/modules/network/vyos/test_vyos_static_route.py:55:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosStaticRouteModule.load_fixtures' method
ERROR: tests/unit/modules/network/vyos/test_vyos_user.py:53:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosUserModule.load_fixtures' method
```

**Comment by GomathiselviS** [comment]
> In terms of the CI errors, the `black` failure was me just misreading the logs last night and I can fix that this evening. The lint errors all are in files I have not touched, so they appear to be in the `main` branch today:
> 
> ```
> ERROR: Found 7 pylint issue(s) which need to be resolved:
> ERROR: plugins/cliconf/vyos.py:171:4: arguments-renamed: Parameter 'newline' has been renamed to 'output' in overridden 'Cliconf.get' method
> ERROR: plugins/cliconf/vyos.py:171:4: arguments-renamed: Parameter 'output' has been renamed to 'newline' in overridden 'Cliconf.get' method
> ERROR: tests/unit/mock/loader.py:49:4: arguments-renamed: Parameter 'file_name' has been renamed to 'path' in overridden 'DictDataLoader._get_file_contents' method
> ERROR: tests/unit/modules/network/vyos/test_vyos_interface.py:89:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosInterfaceModule.load_fixtures' method
> ERROR: tests/unit/modules/network/vyos/test_vyos_ping.py:52:27: use-maxsplit-arg: Use str(command).split(' | ', maxsplit=1)[0] instead
> ERROR: tests/unit/modules/network/vyos/test_vyos_static_route.py:55:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosStaticRouteModule.load_fixtures' method
> ERROR: tests/unit/modules/network/vyos/test_vyos_user.py:53:4: arguments-renamed: Parameter 'filename' has been renamed to 'transport' in overridden 'TestVyosUserModule.load_fixtures' method
> ```

The pylint errors should be fixed by [204](https://github.com/ansible-collections/vyos.vyos/pull/204). The black/lint errors can be fixed by executing `tox -eblack` and `tox -elinters' .

File: plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py (hunk 1 above)
Comment (GomathiselviS) [reply plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py:845]: I understand your reason for this change. But we prefer to have it the way it was because we follow a similar naming convention across all our modules in the vyos collection.

**Comment by sdwilsh** [comment]
Is there any chance this can get reviewed and merged before the next release?

**Review by GomathiselviS — APPROVED** [review]
LGTM

**Comment by GomathiselviS** [comment]
recheck
